### PR TITLE
Adding FuXi global model

### PR DESCRIPTION
**Pull request by NickGeneva**
# Earth2Studio Pull Request

## Description

(Finally) adds FuXi cascade weather model. One of the more SoTa models presently, similar to pangu requires a set off onnx models. Except the models are used for each 5 day interval, not different time step sizes

To verify it runs exactly like the source repo I ran 10 time-steps of each model from the same initial condition. Results from earth2studio using deterministic and the source repo script are the same.

Some addition misc stuff in this repo:

- Added few checks in other model tests to catch an error where time coordinate was not updated
- Added option to not hash names of hf files... this is related to the onnx files (https://github.com/NVIDIA/earth2studio/issues/46)

r500_short
r500_medium
r500_long

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.

### CODE REVIEW

File: earth2studio/models/px/fuxi.py [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+
+    @classmethod
+    def load_model(
+        cls,
+        package: Package,
+    ) -> PrognosticModel:
+        """Load prognostic from package"""
+        # Ghetto at the moment because NGC files are zipped. This will download zip and
+        # unpack them then give the cached folder location from which we can then
+        # access the needed files.
```
Comment (dallasfoster) [inline earth2studio/models/px/fuxi.py]: Left-over comments?

Comment (NickGeneva) [reply earth2studio/models/px/fuxi.py]: Yep! Good spot.

**Review by dallasfoster — APPROVED** [review]

**Comment by NickGeneva** [comment]
For Documentation purposes, this is the script I used for reference from the original repo:

`20231012-06_input_netcdf.nc` was provided from the [Zenodo data](https://zenodo.org/records/10401602) repo in the samples folder.

```
import torch
import argparse
import os
import time 
import numpy as np
import xarray as xr
import pandas as pd
import onnxruntime as ort
from datetime import datetime

from earth2studio.io import ZarrBackend

ort.set_default_logger_severity(3)

def time_encoding(init_time, total_step, freq=6):
    init_time = np.array([init_time])
    tembs = []
    for i in range(total_step):
        hours = np.array([pd.Timedelta(hours=t*freq) for t in [i-1, i, i+1]])
        times = init_time[:, None] + hours[None]
        times = [pd.Period(t, 'H') for t in times.reshape(-1)]
        times = [(p.day_of_year/366, p.hour/24) for p in times]
        temb = np.array(times, dtype=np.float32)
        temb = np.concatenate([np.sin(temb), np.cos(temb)], axis=-1)
        temb = temb.reshape(1, -1)
        tembs.append(temb)
    return np.stack(tembs)



def load_model(model_name):
    # Set the behavier of onnxruntime
    options = ort.SessionOptions()
    options.enable_cpu_mem_arena=False
    options.enable_mem_pattern = False
    options.enable_mem_reuse = False
    # Increase the number for faster inference and more memory consumption
    options.intra_op_num_threads = 1
    cuda_provider_options = {'arena_extend_strategy':'kSameAsRequested',}

    session = ort.InferenceSession(
        model_name,  
        sess_options=options, 
        providers=[('CUDAExecutionProvider', cuda_provider_options)]
    )
    return session

from earth2studio.models.px.fuxi import VARIABLES

def run_inference(model_dir, data, num_steps, save_dir=""):

    total_step = sum(num_steps)
    init_time = pd.to_datetime(data.time.values[-1])
    tembs = time_encoding(init_time, total_step)

    print(f'init_time: {init_time.strftime(("%Y%m%d-%H"))}')
    print(f'latitude: {data.lat.values[0]} ~ {data.lat.values[-1]}')
    
    assert data.lat.values[0] == 90
    assert data.lat.values[-1] == -90

    import xarray as xr
    data = xr.open_dataarray("fuxi.nc")
    data = data.to_numpy()
    input = data[None,...]
    input = input.astype(np.single)
    print(input.dtype)

    # input = data.values[None]
    print(f'input: {input.shape}, {input.min():.2f} ~ {input.max():.2f}')
    print(f'tembs: {tembs.shape}, {tembs.mean():.4f}')

    stages = ['short', 'medium', 'long']
    step = 0

    io = ZarrBackend("outputs/fuxi_original.zarr")

    total_coords = {
        "time": np.array([np.datetime64(datetime(2023,10,12,6))]),
        "lead_time": np.array([np.timedelta64(i*6, "h") for i in range(0, sum(num_steps)+1)]),
        "lat": np.linspace(90, -90, 721, endpoint=True),
        "lon": np.linspace(0, 360, 1440, endpoint=False),
    }
    io.add_array(total_coords, VARIABLES)
    # Save initial state
    data = torch.Tensor(input[:,-1])
    data = torch.split(data, 1, dim=1)
    out_coords = total_coords.copy()
    out_coords['lead_time'] = np.array([np.timedelta64(0, "h")])
    io.write(data, out_coords, VARIABLES)

    for i, num_step in enumerate(num_steps):
        stage = stages[i]
        start = time.perf_counter()
        model_name = os.path.join(model_dir, f"{stage}.onnx")
        print(f'Load model from {model_name} ...')        
        session = load_model(model_name)
        load_time = time.perf_counter() - start
        print(f'Load model take {load_time:.2f} sec')

        print(f'Inference {stage} ...')
        start = time.perf_counter()

        for j in range(0, num_step):
            temb = tembs[step]
            new_input, = session.run(None, {'input': input, 'temb': temb})
            output = torch.Tensor(new_input[:, -1]) 
            data = torch.split(output, 1, dim=1)
            
            print(len(data), data[0].shape)
            
            out_coords = total_coords.copy()
            out_coords['lead_time'] = np.array([np.timedelta64((i*num_step + (j+1))*6, "h")])
            
            io.write(data, out_coords, VARIABLES)
            
            print(f'stage: {i}, step: {step+1:02d}, output: {output.min():.2f} {output.max():.2f}')
            input = new_input
            step += 1

        run_time = time.perf_counter() - start
        print(f'Inference {stage} take {run_time:.2f}')

        if step > total_step:
            break


    
if __name__ == "__main__":
    data = xr.open_dataarray("20231012-06_input_netcdf.nc")
    model_dir = "./models"
    output_dir = "./outputs/original"
    run_inference(model_dir, data, [20,20,20], output_dir)
    
```

**Comment by NickGeneva** [comment]
This is the script I used to test e2studio where `fuxi.nc` is the same input net CDF but with renamed coordinates to match e2studio standards:

```
model = FuXi.load_model(FuXi.load_default_package())

import os
if not os.path.exists("fuxi.nc"):
    # Create the data source
    data = CustomDataSource()
    datasource_to_file("fuxi.nc", data, time=["2023-10-12 06:00:00"],
            variable=model.input_coords["variable"],
            lead_time=model.input_coords["lead_time"])

data = DataArrayFile("fuxi.nc")
# Create the IO handler, store in memory
io = ZarrBackend("outputs/fuxi_e2studio.zarr")

# # %%
# # Execute the Workflow
# # --------------------
# # With all components initialized, running the workflow is a single line of Python code.
# # Workflow will return the provided IO object back to the user, which can be used to
# # then post process. Some have additional APIs that can be handy for post-processing or
# # saving to file. Check the API docs for more information.
# #
# # For the forecast we will predict for two days (these will get executed as a batch) for
# # 20 forecast steps which is 5 days.

# # %%
import earth2studio.run as run

nsteps = 60
io = run.deterministic(["2023-10-12 06:00:00"], nsteps, model, data, io)

print(io.root.tree())
```

**Comment by NickGeneva** [comment]
Full three model roll out to 15 days
r500_full

**Pull request merged**